### PR TITLE
For MSVC use alloca() instead of VLA

### LINKED ISSUE #20: Variable-length arrays are not compatible with MSVC
In the category "Microsoft compilers suck", such as previously seen in #14, the use of variable-length arrays is not supported by Microsoft compilers. This means even the `ref` implementation doesn't work on Windows.

### PR DESCRIPTION
**Pull request by bwesterb**
Closes #20

@thomwiggers

### CODE REVIEW

**Comment by thomwiggers** [comment]
Was this tested? AFAICT MSVC doesn't have alloca, but it does have _alloca (deprecated) in malloc.h (for maximum confusion).

**Comment by sfluhrer** [comment]
I'm not certain that allocating variable length automatic arrays (either with VLAs or using alloca) is the Right Thing.  Two examples:
- haraka_S_absorb - the rate is always 32 - why do we need a variable length array?
- thash (the robust versions) - currently, we allocate the full array, generate the full mask, and xor it in.  An alternative approach would be generate the mask one block at a time, and process it incrementally - see https://github.com/sphincs/parallel-sphincsplus/blob/main/shake256_robust.cpp for an example of that being done.

File: ref/utils.h [hunk 1]
```diff
@@ -6,6 +6,14 @@
 #include "context.h"
 
 
+/* To support MSVC use alloca() instead of VLAs. See #20. */
+#ifdef _MSC_VER
+# include <alloca.h>
+# define SPX_VLA(__t,__x,__s) __t *__x = (__t*)alloca((__s)*sizeof(__t))
```
Comment (thomwiggers) [inline ref/utils.h:12]:
```suggestion
/* MSVC defines _alloca in malloc.h */
# include <malloc.h>
/* Note: _malloca(), which is recommended over deprecated _alloca,
   requires that you call _freea(). So we stick with _alloca */ 
# define SPX_VLA(__t,__x,__s) __t *__x = (__t*)_alloca((__s)*sizeof(__t))
```

This is required for Windows support as tested on Godbolt.

**Review by thomwiggers — CHANGES_REQUESTED** [review]

**Comment by thomwiggers** [comment]
Even though this perhaps isn't at all the best way to solve this, it would be great if we can actually merge some form of this PR so that we can proceed with updating PQClean (and OQS downstream).

**Comment by bwesterb** [comment]
@sfluhrer I want to propose to merge a rebased version of this patch together with Thom's suggestion for now. Then, later, we can get rid of the variable sized arrays alltogether.

**Review by sfluhrer — APPROVED** [review]

**Comment by bwesterb** [comment]
@thomwiggers Rebased. If builds pass, please give it a double-check.

**Comment by thomwiggers** [comment]
Seems to compile just fine. There's some more work needed before this repo will produce useful binaries (mostly because the test binaries here use OpenSSL and the random number generator isn't set up for Windows) -- but I'm not getting VLA problems here.

**Pull request merged**